kinova_driver/src/scripts/pid_controller.py

```diff
@@ -0,0 +1,293 @@
+#!/usr/bin/env python
+
+"""
+Author: Stephen Hansen
+Adapted from Andrea Bajcsy
+https://github.com/abajcsy/iact_control/
+"""
+
+from enum import Enum
+import sys
+import time
+
+import numpy as np
+from numpy import array
+import roslib
+import rospy
+import kinova_msgs.msg
+import geometry_msgs.msg
+import std_msgs.msg
+import sensor_msgs.msg
+import kinova_msgs.srv
+
+from trajectory_msgs.msg import JointTrajectory, JointTrajectoryPoint
+
+import pid
+import ros_utils
+import actionlib
+
+roslib.load_manifest('kinova_demo')
+
+PREFIX = 'j2s7s300_driver'
+
+GOAL_EPSILON = 0.005 #if all joints are less than epsilon at goal, then complete, goal is achieved
+START_EPSILON = 0.1 #if all joints are less than epsilon at start, the trajectory can begin (prevents jumping)
+START_MAX_DIST = 0.5 #if ANY joint is greater than start max distance, throw an error
+
+MAX_CMD_VEL = 35.0 #Maximum commanded velocity
+
+class PIDController(object):
+    """
+    This class represents a node that moves the Jaco with PID control.
+    The joint velocities are computed as:
+
+        V = -K_p(e) - K_d(e_dot) - K_i*Integral(e)
+    where:
+        e = (target_joint configuration) - (current joint configuration)
+        e_dot = derivative of error
+        K_p = accounts for present values of position error
+        K_i = accounts for past values of error, accumulates error over time
+        K_d = accounts for possible future trends of error, based on current rate of change
+
+    Subscribes to:
+        /j2s7s300_driver/out/joint_angles	- Jaco sensed joint angles
+        /j2s7s300_driver/out/joint_torques	- Jaco sensed joint torques
+
+    Publishes to:
+        /j2s7s300_driver/in/joint_velocity	- Jaco commanded joint velocities
+
+    Required parameters:
+        p_gain, i_gain, d_gain    - gain terms for the PID controller
+        sim_flag 				  - flag for if in simulation or not
+    """
+
+    def __init__(self):
+        """
+        Setup of the ROS node. Publishing computed torques happens at 100Hz.
+        """
+
+        self.reached_start = False
+        self.reached_goal = False
+        self.last_dof = None
+        self.target_index = 0
+        self.step_size = 1
+        self.time_points = None
+
+        # ----- Controller Setup ----- #
+
+        # stores maximum COMMANDED joint torques
+        self.max_cmd = MAX_CMD_VEL * np.eye(7)
+        # stores current COMMANDED joint torques
+        self.cmd = np.eye(7)
+        # stores current joint MEASURED joint torques
+        self.joint_torques = np.zeros((7, 1))
+
+        # P, I, D gains
+        p_gain = 60.0
```

Comment:
Can you elaborate about your concern Martine? I see no issue with setting PID gains here.